open_spiel/python/pytorch/ppo.py

```diff
@@ -0,0 +1,345 @@
+# Note: code adapted (with permission) from https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo.py and https://github.com/vwxyzjn/ppo-implementation-details/blob/main/ppo_atari.py
+
+import time
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.distributions.categorical import Categorical
+
+from open_spiel.python.rl_agent import StepOutput
+
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+class PPOAgent(nn.Module):
+    def __init__(self, num_actions, observation_shape, device):
+        super().__init__()
+        self.critic = nn.Sequential(
+            layer_init(nn.Linear(np.array(observation_shape).prod(), 64)),
+            nn.Tanh(),
+            layer_init(nn.Linear(64, 64)),
+            nn.Tanh(),
+            layer_init(nn.Linear(64, 1), std=1.0),
+        )
+        self.actor = nn.Sequential(
+            layer_init(nn.Linear(np.array(observation_shape).prod(), 64)),
+            nn.Tanh(),
+            layer_init(nn.Linear(64, 64)),
+            nn.Tanh(),
+            layer_init(nn.Linear(64, num_actions), std=0.01),
+        )
+        self.device = device
+        self.num_actions = num_actions
+
+    def get_value(self, x):
+        return self.critic(x)
+
+    def get_action_and_value(self, x, legal_actions_mask=None, action=None):
+        if legal_actions_mask is None:
+            # All valid
+            legal_actions_mask = torch.ones((len(x), self.num_actions)).bool()
+
+        # Fill with invalids
+        INVALID_ACTION_PENALTY = -1e6
+        logits = torch.full((len(x), self.num_actions), INVALID_ACTION_PENALTY).to(self.device)
```

Comment:
It's undesirable to create this `logits` full of `-1e6` every time during action sampling. It's better to save this as part of the module's buffer and reuse. See

https://github.com/vwxyzjn/gym-microrts/blob/e4e11405a36044eab49d4cd2c2ed084b019bb999/experiments/ppo_gridnet.py#L199
https://github.com/vwxyzjn/gym-microrts/blob/e4e11405a36044eab49d4cd2c2ed084b019bb999/experiments/ppo_gridnet.py#L211

Reply:
Ah, cool! Thanks for the pointer